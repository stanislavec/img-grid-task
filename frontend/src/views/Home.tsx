import React, { useMemo } from 'react';
import GridContainer from 'components/GridContainer';
import Loader from 'components/Loader';
import usePhotoContext from 'context/useImageContext';
import useBodyScroll from 'hooks/useBodyScroll';
import { IPhoto } from 'types';

import scss from './Home.module.scss';

type IGridContainerItem = {
  photo: IPhoto;
  onClick(): void;
};

function GridContainerItem(props: IGridContainerItem) {
  const { favorites } = usePhotoContext();
  const { onClick, photo } = props;

  const isFavorite = useMemo(
    () => favorites.some((favorite) => favorite.id === photo.id),
    [favorites, photo.id],
  );

  return (
    <div className={scss.gridItem} onClick={onClick}>
      {isFavorite && <div className={scss.favorite}>❤️</div>}
      <img src={photo.download_url} alt={photo.url} />
    </div>
  );
}

function Home() {
  const { photos, fetchNext, loading, addToFavorite } = usePhotoContext();
  useBodyScroll(fetchNext);

  return (
    <>
      <GridContainer
        isCheckedToBeEmpty={!loading}
        items={photos}
        renderItem={(photo) => (
          <GridContainerItem photo={photo} onClick={() => addToFavorite(photo)} />
        )}
        getKey={(photo) => photo.id}
      ></GridContainer>
      {loading && <Loader />}
    </>
  );
}

export default Home;
