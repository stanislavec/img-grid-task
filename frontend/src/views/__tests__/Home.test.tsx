import { fireEvent, render } from '@testing-library/react';
import usePhotoContext from 'context/useImageContext';
import Home from '../Home';
import PHOTOS from './__mocks__/photo';

jest.mock('context/useImageContext');

const mockUsePhotoContext = usePhotoContext as jest.MockedFunction<typeof usePhotoContext>;

describe('Home', () => {
  beforeEach(() => jest.clearAllMocks());

  it('is defined', () => {
    mockUsePhotoContext.mockReturnValue({
      photos: [],
      fetchNext: () => null,
      loading: false,
      addToFavorite: () => null,
      removeFromFavorite: () => null,
      favorites: [],
    });
    const elem = render(<Home />);
    expect(elem.baseElement).toBeDefined();
  });

  it('matches snapshot on loading', () => {
    mockUsePhotoContext.mockReturnValue({
      photos: [],
      fetchNext: () => null,
      loading: true,
      addToFavorite: () => null,
      removeFromFavorite: () => null,
      favorites: [],
    });

    const elem = render(<Home />);
    expect(elem.container).toMatchSnapshot();
  });

  it('matches snapshot on loaded with photos', () => {
    mockUsePhotoContext.mockReturnValue({
      photos: PHOTOS,
      fetchNext: () => null,
      loading: false,
      addToFavorite: () => null,
      removeFromFavorite: () => null,
      favorites: [],
    });

    const elem = render(<Home />);
    expect(elem.container).toMatchSnapshot();
  });

  it('matches snapshot on empty photos collection', () => {
    mockUsePhotoContext.mockReturnValue({
      photos: [],
      fetchNext: () => null,
      loading: false,
      addToFavorite: () => null,
      removeFromFavorite: () => null,
      favorites: [],
    });

    const elem = render(<Home />);
    expect(elem.container).toMatchSnapshot();
  });

  it('handles click to photo', () => {
    const addToFavorite = jest.fn();
    mockUsePhotoContext.mockReturnValue({
      photos: PHOTOS,
      fetchNext: () => null,
      loading: false,
      addToFavorite,
      removeFromFavorite: () => null,
      favorites: [],
    });

    const elem = render(<Home />);
    const image = elem.container.querySelector('img');

    if (!image) throw Error();

    fireEvent.click(image);
    expect(addToFavorite).toBeCalledTimes(1);
  });

  it('matches snapshot with favorites', () => {
    mockUsePhotoContext.mockReturnValue({
      photos: PHOTOS,
      fetchNext: () => null,
      loading: false,
      addToFavorite: () => null,
      removeFromFavorite: () => null,
      favorites: PHOTOS.slice(0, 2),
    });

    const elem = render(<Home />);

    expect(elem.container).toMatchSnapshot();
  });

  it('fires fetch photos on scroll down reached', () => {
    const fetchNext = jest.fn();
    mockUsePhotoContext.mockReturnValue({
      photos: PHOTOS,
      fetchNext,
      loading: false,
      addToFavorite: () => null,
      removeFromFavorite: () => null,
      favorites: [],
    });

    render(<Home />);

    expect(fetchNext).toBeCalledTimes(0);

    fireEvent.scroll(document, { target: { scrollY: 10000 } });

    expect(fetchNext).toBeCalledTimes(1);

    fireEvent.scroll(document, { target: { scrollY: 20000 } });

    expect(fetchNext).toBeCalledTimes(2);
  });
});
