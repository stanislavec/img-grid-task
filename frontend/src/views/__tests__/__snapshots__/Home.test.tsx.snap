// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches snapshot on empty photos collection 1`] = `
<div>
  <div
    class="empty"
  >
    No data...
  </div>
</div>
`;

exports[`Home matches snapshot on loaded with photos 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/yC-Yzbqy7PY"
        src="https://picsum.photos/id/0/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/LNRyGwIJr5c"
        src="https://picsum.photos/id/1/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/N7XodRrbzS0"
        src="https://picsum.photos/id/2/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/Dl6jeyfihLk"
        src="https://picsum.photos/id/3/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/y83Je1OC6Wc"
        src="https://picsum.photos/id/4/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/LF8gK8-HGSg"
        src="https://picsum.photos/id/5/5000/3334"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/tAKXap853rY"
        src="https://picsum.photos/id/6/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/BbQLHCpVUqA"
        src="https://picsum.photos/id/7/4728/3168"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/xII7efH1G6o"
        src="https://picsum.photos/id/8/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/ABDTiLqDhJA"
        src="https://picsum.photos/id/9/5000/3269"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/6J--NXulQCs"
        src="https://picsum.photos/id/10/2500/1667"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/Cm7oKel-X2Q"
        src="https://picsum.photos/id/11/2500/1667"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/I_9ILwtsl_k"
        src="https://picsum.photos/id/12/2500/1667"
      />
    </div>
  </div>
</div>
`;

exports[`Home matches snapshot on loading 1`] = `
<div>
  <div
    class="container"
  />
  <div
    class="row"
  >
    <div
      class="loader"
    />
  </div>
</div>
`;

exports[`Home matches snapshot with favorites 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="gridItem"
    >
      <div
        class="favorite"
      >
        ❤️
      </div>
      <img
        alt="https://unsplash.com/photos/yC-Yzbqy7PY"
        src="https://picsum.photos/id/0/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <div
        class="favorite"
      >
        ❤️
      </div>
      <img
        alt="https://unsplash.com/photos/LNRyGwIJr5c"
        src="https://picsum.photos/id/1/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/N7XodRrbzS0"
        src="https://picsum.photos/id/2/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/Dl6jeyfihLk"
        src="https://picsum.photos/id/3/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/y83Je1OC6Wc"
        src="https://picsum.photos/id/4/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/LF8gK8-HGSg"
        src="https://picsum.photos/id/5/5000/3334"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/tAKXap853rY"
        src="https://picsum.photos/id/6/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/BbQLHCpVUqA"
        src="https://picsum.photos/id/7/4728/3168"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/xII7efH1G6o"
        src="https://picsum.photos/id/8/5000/3333"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/ABDTiLqDhJA"
        src="https://picsum.photos/id/9/5000/3269"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/6J--NXulQCs"
        src="https://picsum.photos/id/10/2500/1667"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/Cm7oKel-X2Q"
        src="https://picsum.photos/id/11/2500/1667"
      />
    </div>
    <div
      class="gridItem"
    >
      <img
        alt="https://unsplash.com/photos/I_9ILwtsl_k"
        src="https://picsum.photos/id/12/2500/1667"
      />
    </div>
  </div>
</div>
`;
